File: programmers/lv2/기출문제/도넛과 막대 그래프.js
const DONUT = 1, STICK = 2, EIGHT = 3;

function solution(edges) {
    var answer = [0, 0, 0, 0];
    
    const nodeStartMap = new Map(), nodeEndMap = new Map();
    edges.forEach(edge => {
        const [start, end] = edge;
        if(!nodeStartMap.has(start)) {
            nodeStartMap.set(start, 0);
        }
        if(!nodeStartMap.has(end)) {
            nodeStartMap.set(end, 0);
        }
        if(!nodeEndMap.has(start)) {
            nodeEndMap.set(start, 0);
        }
        if(!nodeEndMap.has(end)) {
            nodeEndMap.set(end, 0);
        }
        nodeStartMap.set(start, nodeStartMap.get(start) + 1);
        nodeEndMap.set(end, nodeEndMap.get(end) + 1);

    })
    
    var startNode = -1;
    for(const node of nodeStartMap.keys()) {
        if(node <= 0) {
            break;
        }
        if(nodeStartMap.get(node) > 1 && nodeEndMap.get(node) === 0) {
            startNode = node;
            answer[0] = startNode;
            break;
        }
    }
    
    var graphNum = nodeStartMap.get(startNode);
    for(const [key, value] of nodeStartMap.entries()) {
        if(key === startNode) {
            continue;
        }
        if(value === 2) {
            answer[EIGHT]++;
            graphNum--;
        } else if(value === 0) {
            answer[STICK]++;
            graphNum--;
        }
    }
    answer[DONUT] += graphNum;
    
    return answer;
}

File: programmers/lv2/기출문제/도넛과 막대 그래프.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        
        Map<Integer, List<Integer>> map = new HashMap<>();
        Set<Integer> inSet = new HashSet<>();
        for(int[] edge: edges) {
            map.putIfAbsent(edge[0], new ArrayList<>());
            map.get(edge[0]).add(edge[1]);
            inSet.add(edge[1]);
        }
        
        for(int key: map.keySet()) {
            if(map.get(key).size() >= 2 && !inSet.contains(key)) {
                answer[0] = key;
            }
        }
        
        Set<Integer> visited = new HashSet<>();
        visited.add(answer[0]);
        label: for(int startIdx: map.get(answer[0])) {
            visited.add(startIdx);
            Queue<Integer> q = new ArrayDeque<>();
            q.offer(startIdx);
            while(!q.isEmpty()) {
                int node = q.poll();
                if(map.get(node) == null) {
                    answer[2]++;
                    continue label;
                }
                for(int idx: map.get(node)) {
                    if(map.get(idx) == null) {
                        answer[2]++;
                        continue label;
                    }
                    if(map.get(idx).size() == 2) {
                        answer[3]++;
                        continue label;
                    }
                    if(!visited.contains(idx)) {
                        visited.add(idx);
                        q.offer(idx);
                    }
                }
            }
            answer[1]++;
        }
        
        return answer;
    }
}
